### ISSUE TITLE
Analyses fail randomly due to IndexError

### ISSUE THREAD

**Issue by hackdna** [refinery-platform/refinery-platform#2993]
* Specific code commit: 137a461
* Version of the web browser and OS: any
* Environment where the error occurred: Vagrant

### Steps to reproduce
Run an analysis using a [test workflow](https://github.com/refinery-platform/galaxy-tools/blob/develop/refinery_test_workflows/Test_workflow__5_steps_without_branching.ga)

### Observed behavior
Analysis completes successfully in Galaxy but fails randomly in Refinery with the following error message in the Celery log:
```
2018-09-07 11:26:27,619 ERROR    celery.worker.job:282 log() - Task analysis_manager.tasks.run_analysis[75d2a63a-c8b0-4bd3-a710-e332bacba1fd] raised unexpected: IndexError('list index out of range',)
Traceback (most recent call last):
  File "/home/vagrant/.virtualenvs/refinery-platform/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/home/vagrant/.virtualenvs/refinery-platform/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
    return self.run(*args, **kwargs)
  File "/vagrant/refinery/analysis_manager/tasks.py", line 302, in run_analysis
    _attach_workflow_outputs(analysis_uuid)
  File "/vagrant/refinery/analysis_manager/tasks.py", line 149, in _attach_workflow_outputs
    analysis.attach_derived_nodes_to_dataset()
  File "/vagrant/refinery/core/models.py", line 1414, in attach_derived_nodes_to_dataset
    self._create_derived_data_file_nodes(graph_with_input_nodes_linked)
  File "/vagrant/refinery/core/models.py", line 1589, in _create_derived_data_file_nodes
    self._get_output_connection_to_analysis_result_mapping()
  File "/vagrant/refinery/core/models.py", line 1508, in _get_output_connection_to_analysis_result_mapping
    )[index]
  File "/home/vagrant/.virtualenvs/refinery-platform/local/lib/python2.7/site-packages/django/db/models/query.py", line 201, in __getitem__
    return list(qs)[0]
IndexError: list index out of range
```

### Expected behavior
No error

### Notes
There is [no default order](https://stackoverflow.com/a/7164126) for items returned by [filter() call](https://github.com/refinery-platform/refinery-platform/blob/137a461f3cd8b03402034bf1e105c68ce3cf3343/refinery/core/models.py#L1487).

![screen shot 2018-09-11 at 11 51 29](https://user-images.githubusercontent.com/452575/45371962-ca280e00-b5b9-11e8-9d43-9569c2f271a6.png)
![screen shot 2018-09-11 at 11 52 54](https://user-images.githubusercontent.com/452575/45371963-ca280e00-b5b9-11e8-8bd4-bf5aaf6ea1ca.png)

Also, this cause outputs to be added incorrectly to the provenance graph in case of analyses with more than one input.

**Comment by scottx611x** [comment]
This error occurs **100% of the time** with my local setup, **~50% of the time** with Ilya's local setup, and **0% of the time** with a CloudMan Cluster and AWS Refinery instance. 

I'm hesitant to write a fix for something that we haven't observed in our production environment. Especially due to the fact that I am able to reproduce this error reliably with my local setup.

**Comment by scottx611x** [comment]
TL;DR - This error occurs **~50% of the time** with Ilya's local setup, **100% of the time** with my local setup, and **0% of the time** with a CloudMan Cluster and AWS Refinery instance. I'm hesitant to write a fix for something that we haven't observed in our production environment. Especially due to the fact that I am able to reproduce this error reliably with my local setup.

### Notes from past conversations:

scott [2:59 PM]
Hey Ilya, I'm unable to reproduce the `IndexError` you saw within my existing tests when I make the order of that filter you mentioned random explicitly ( `.order_by('?')`).

`core.tests.AnalysisTests.test___get_output_connection_to_analysis_result_mapping` will fail when I order randomly, but not due to an `IndexError`. Just because the result differs in ordering now from the expected.

I really think that this is due to some slight difference in Galaxy setups. For example we're interacting with Galaxy quite a bit in `tool_manager.models.WorkflowTool.create_analysis_output_node_connections` whose resulting created objects is what the filter you mentioned is actually drawing from.

Testing with my local galaxy 17.05 instance I'm able to reproduce the `IndexError`.  But I'm hesitant still to try and make an actual fix for this specific scenario if this isn't what we're experiencing on our CloudMan Cluster.

The following patch allows me to run the Analysis successfully, but **fails to add any derived results to the DataSet**.
Also, **please take note how my local 17.05 instance yields duplicate values in `retained_datasets` which is not happening on our CloudMan Galaxy Instance**.
```
diff --git a/refinery/core/models.py b/refinery/core/models.py
index 88ee7f232..b6b3145e4 100644
--- a/refinery/core/models.py
+++ b/refinery/core/models.py
@@ -1487,7 +1487,7 @@ class Analysis(OwnableResource):
         output_node_connections = AnalysisNodeConnection.objects.filter(
             analysis=self,
             direction=OUTPUT_CONNECTION
-        )
+        ).order_by("-step")
         # Fetch the distinct file names from our output
         # AnalysisNodeConnections for this Analysis construct a dict
         # mapping the unique file names to a list of AnalysisNodeConnections
diff --git a/refinery/tool_manager/models.py b/refinery/tool_manager/models.py
index efc4ea5b6..1ff0ac4fe 100644
--- a/refinery/tool_manager/models.py
+++ b/refinery/tool_manager/models.py
@@ -1099,7 +1099,10 @@ class WorkflowTool(Tool):
             if not galaxy_dataset["purged"] and
             self._get_workflow_step(galaxy_dataset) > 0
         ]
-        return retained_datasets
+
+        # My local Galaxy 17.05 yields a duplicate for each entry in
+        # retained values
+        return {d['type_id']: d for d in retained_datasets}.values()
 
     def _get_exposed_galaxy_datasets(self):
         """
```

I would be more comfortable making a change if we can reproduce this on our production setup or a local Galaxy setup that truly emulates our production one.

ilya [3:40 PM]
The code in question is brittle because it relies on the order of items in the list of AnalysisNodeConnections to look up values in the list of AnalysisResults. That needs to be fixed regardless of any additional issues with Galaxy.

ilya [11:53 AM]
Just ran into this problem again

---

Scott - "I've just run ~10 more analyses [23 Total] and still have a **100% Failure rate** (all with that `IndexError` in `_get_output_connection_to_analysis_result_mapping`) so far with my local environment which is a huge discrepancy between what you're seeing, and what we see in our production cluster."

![screen shot 2018-09-11 at 12 36 59 pm](https://user-images.githubusercontent.com/5629547/45386091-78df4500-b5e0-11e8-955d-3e40a25cf1f3.png)

**Comment by hackdna** [comment]
The code in question is broken by design. Different error rates are simply due to random order of items returned by the filter() function. It's just a matter of time before this error occurs in production. Also, it may have already caused some analysis outputs to be attached to wrong inputs.

**Comment by scottx611x** [comment]
After taking the time and testing this with a CloudMan the error is 100% reproducible and isolated to the `Test workflow: 5 steps without branching` workflow. 

@hackdna Your **50% success rate** was because you were also running the `Test workflow: SPP analog` workflow locally which doesn't exhibit this issue.
My **100% Failure rate** was due to the fact that I was only running `Test workflow: 5 steps without branching` locally.

What the error stems from is the fact that we hadn't updated the Rename DataSet field properly for the `Test workflow: 5 steps without branching` workflow to give a more descriptive/unique output name for the outputs that we are bringing back into Refinery like we have been with other workflows. We saw the `IndexError` because each output file name was `Output file.txt`.

When this is done properly the order of that `filter()` shouldn't matter.

You can try this out locally with the following updated version of `Test_workflow__5_steps_without_branching.ga` 
(Please note I had to add the `.txt` extension to upload to Github and that should be removed before adding to Galaxy)
[Test_workflow__5_steps_without_branching.ga.txt](https://github.com/refinery-platform/refinery-platform/files/2375859/Test_workflow__5_steps_without_branching.ga.txt)

**Comment by hackdna** [comment]
I did not use Test workflow: SPP analog. The IndexError is due to the value of the `index` variable derived from a randomly ordered list of one size used for another randomly ordered list of a different size. Renaming of inputs is only going to mask this problem.

**Comment by scottx611x** [comment]
@hackdna are you able to reproduce the failure condition with the Workflow that I've provided?

This error will not occur (nor do we care about the ordering of the mentioned filter) with properly annotated workflows. Specifically the `Test_workflow__5_steps_without_branching.ga` is the only one of our workflows that will exhibit this issue, Could you please  test with your local setup and confirm?

**Comment by hackdna** [comment]
Updating workflow annotations is useful but it does not fix fundamental problems with the code in _get_output_connection_to_analysis_result_mapping().

**Comment by scottx611x** [comment]
@hackdna That didn't answer my question. 

When we properly annotate our workflows (as we should be doing to provide more meaningful names to the derived results of our workflow runs Ref https://github.com/refinery-platform/refinery-platform/issues/2373) the ordering of the filter that you have mentioned does not matter.

I believe that I have done my part to dive deeper and investigate the root cause of the issue that you're experiencing. It would be helpful if you could confirm that this behavior does not exist with the workflow I've provided you.

If you still feel that this issue is critical please feel free to present your findings in our next team meeting.

**Comment by hackdna** [comment]
The function in question `_get_output_connection_to_analysis_result_mapping()` is also causing random [Travis build fails](https://travis-ci.org/refinery-platform/refinery-platform/builds/436321330):
```
FAIL: test___get_output_connection_to_analysis_result_mapping (core.tests.AnalysisTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/refinery-platform/refinery-platform/refinery/core/tests.py", line 718, in test___get_output_connection_to_analysis_result_mapping
    (self.analysis_node_connection_a, analysis_result_1)
AssertionError: Lists differ: [(<AnalysisNodeConnection: out... != [(<AnalysisNodeConnection: out...
First differing element 0:
(<AnalysisNodeConnection: out: 1_ (True) None>, <AnalysisResult: test_node.txt <-> 66baa46c-04bf-407e-b010-054299674cbd>)
(<AnalysisNodeConnection: out: 3_ (True) None>, <AnalysisResult: test_node.txt <-> 66baa46c-04bf-407e-b010-054299674cbd>)
- [(<AnalysisNodeConnection: out: 1_ (True) None>,
?                                 ^
+ [(<AnalysisNodeConnection: out: 3_ (True) None>,
?                                 ^
    <AnalysisResult: test_node.txt <-> 66baa46c-04bf-407e-b010-054299674cbd>),
   (<AnalysisNodeConnection: out: 2_ (False) None>, None),
-  (<AnalysisNodeConnection: out: 3_ (True) None>,
?                                 ^
+  (<AnalysisNodeConnection: out: 1_ (True) None>,
?                                 ^
    <AnalysisResult: test_node.txt <-> 66baa46c-04bf-407e-b010-054299674cbd>)]
```

**Comment by hackdna** [comment]
A Travis build just failed again due to this issue:
https://travis-ci.org/refinery-platform/refinery-platform/builds/563759322#L6078
```
FAIL: test___get_output_connection_to_analysis_result_mapping (core.test_models.AnalysisTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/refinery-platform/refinery-platform/refinery/core/test_models.py", line 368, in test___get_output_connection_to_analysis_result_mapping
    (self.analysis_node_connection_a, analysis_result_1)
AssertionError: Lists differ: [(<AnalysisNodeConnection: out... != [(<AnalysisNodeConnection: out...
First differing element 0:
(<AnalysisNodeConnection: out: 1_ (True) None>, <AnalysisResult: test_node.txt <-> c7647b1d-8050-4c6e-a4bb-08c58783e01b>)
(<AnalysisNodeConnection: out: 3_ (True) None>, <AnalysisResult: test_node.txt <-> c7647b1d-8050-4c6e-a4bb-08c58783e01b>)
- [(<AnalysisNodeConnection: out: 1_ (True) None>,
?                                 ^
+ [(<AnalysisNodeConnection: out: 3_ (True) None>,
?                                 ^
    <AnalysisResult: test_node.txt <-> c7647b1d-8050-4c6e-a4bb-08c58783e01b>),
   (<AnalysisNodeConnection: out: 2_ (False) None>, None),
-  (<AnalysisNodeConnection: out: 3_ (True) None>,
?                                 ^
+  (<AnalysisNodeConnection: out: 1_ (True) None>,
?                                 ^
    <AnalysisResult: test_node.txt <-> c7647b1d-8050-4c6e-a4bb-08c58783e01b>)]
```
